### ISSUE TITLE
Stubs and std::unique_ptr

### ISSUE THREAD

**Issue by Artoymyp** [eranpeer/FakeIt#40]
How should I stub methods, that return `unique_ptr<T>`? 
If I try it straightforward way, the code will not compile:

``` C++
class Product{};
class IFactory{
public:
    virtual std::unique_ptr<Product>Create() = 0;
};
TEST_CLASS(UnitTest)
{
public:
    TEST_METHOD(TestMethod1)
    {   
        Mock<IFactory> factory;
        When(Method(factory, Create)).Return(std::make_unique<Product>());
        auto a = factory.get().Create();
        Verify(Method(factory, Create)).Once();
    }
};
```

error C2280: 'std::unique_ptr<Product,std::default_delete<_Ty>>::unique_ptr(const std::unique_ptr<_Ty,std::default_delete<_Ty>> &)' : attempting to reference a deleted function.

Seems like the error is caused by an inavailability of copy-constructor. std::shared_ptr version is obviously working fine.

I also have tried partially implementing IFactory to no result:

``` C++
class IFactoryMock : public IFactory{
public:
    virtual Product* CreateRaw() = 0;
    std::unique_ptr<Product> Create(){
        return std::unique_ptr<Product>(CreateRaw());
    }
};

TEST_CLASS(UnitTest)
{
public:
    TEST_METHOD(TestMethod2)
    {
        Mock<IFactoryMock> factory;
        When(Method(factory, CreateRaw)).Return(new Product());
        auto a = factory.get().Create();
        Verify(Method(factory, CreateRaw)).Once();
    }
};
```

Result Message: Assert failed. Unexpected method invocation: unknown()
  An unmocked method was invoked. All used virtual methods must be stubbed!

**Comment by eranpeer** [comment]
When mocking a class all virtual methods are replaced with a default implementation that throws this exception.
Here you stubbed the CreateRaw method but invoked the Create method.
The error message at runtime is correct. You must stub Create method before invoking it.

The other thing you can do is to "spy" an existing object. This way you can stub only part of the methods. For example:
```c++
class IFactoryMock : public IFactory{
public:
    virtual Product* CreateRaw() { return nullptr; };
    std::unique_ptr<Product> Create(){
        return std::unique_ptr<Product>(CreateRaw());
    }
};

void testMethod()
{
    IFactoryMock factory;
    Mock<IFactoryMock> spy(factory);
    When(Method(spy, CreateRaw)).Return(new Product());
    auto a = spy.get().Create();
    Verify(Method(spy, CreateRaw)).Once();
}
```

**Comment by eranpeer** [comment]
And yes you are right. The problem is caused by the fact that the unique_ptr is not copy-constructible.
I will try to fix this one. Maybe I will be able to find a workaround.

**Comment by amerry** [comment]
You can use Do() for this:

```cpp
When(Method(factory, Create)).Do([]{ return std::make_unique<Product>(); });
```

When using unique_ptr, you inherently need to create a new one for every call - using the lambda allows you to do that.

**Comment by FranckRJ** [comment]
I guess this is fixed by #282, which is in FakeIt 2.3.1.
